File: ios_app/iphone/UI/Summary/SummaryLineChartView.swift
import SwiftUI
import shared

struct SummaryLineChartView: View {
    
    let goalsUi: [SummaryVm.GoalUi]
    let daysBarsUi: [DayBarsUi]
    
    var body: some View {
        VmView({
            SummaryLineChartVm(
                goalsUi: goalsUi,
                initialDaysBarsUi: daysBarsUi
            )
        }) { vm, state in
            SummaryLineChartViewInner(
                vm: vm,
                state: state
            )
        }
    }
}

private struct SummaryLineChartViewInner: View {
    
    let vm: SummaryLineChartVm
    let state: SummaryLineChartVm.State
    
    var body: some View {
        
        VStack {
            
            // Time unit selector
            TimeUnitSelector(
                currentUnit: state.timeUnit,
                onUnitSelected: { unit in
                    vm.setTimeUnit(unit: unit)
                }
            )
            
            // Line Chart
            if !state.linesData.isEmpty {
                LineChartView(
                    linesData: state.linesData,
                    maxSeconds: state.maxSecondsOverall,
                    xLabels: state.xLabels
                )
                .frame(height: 220)
                .padding(.horizontal, 8)
                .padding(.vertical, 16)
            } else {
                VStack {
                    Text("Select activities to display")
                        .foregroundColor(.secondary)
                        .font(.system(size: 14))
                }
                .frame(height: 200)
            }
            
            // Goal selection
            Text("Select Activities")
                .font(.system(size: 14, weight: .medium))
                .foregroundColor(Color(.label))
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 16)
                .padding(.vertical, 8)
            
            GoalSelectionList(
                selections: state.goalSelections,
                onToggle: { goalId in
                    vm.toggleGoalSelection(goalId: goalId)
                }
            )
            
            Spacer()
                .frame(height: 20)
        }
        .background(.background)
    }
}

private struct TimeUnitSelector: View {
    
    let currentUnit: SummaryLineChartVm.TimeUnit
    let onUnitSelected: (SummaryLineChartVm.TimeUnit) -> Void
    
    var body: some View {
        HStack {
            ForEach(SummaryLineChartVm.TimeUnit.entries, id: \.self) { unit in
                let isSelected = unit == currentUnit
                
                Button(action: {
                    onUnitSelected(unit)
                }) {
                    Text(unit.title)
                        .font(.system(size: 14, weight: isSelected ? .medium : .regular))
                        .foregroundColor(isSelected ? .white : Color(.label))
                        .padding(.horizontal, 16)
                        .padding(.vertical, 8)
                        .background(
                            RoundedRectangle(cornerRadius: 8, style: .continuous)
                                .fill(isSelected ? Color.blue : Color.clear)
                        )
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }
}

private struct LineChartView: View {
    
    let linesData: [SummaryLineChartVm.LineData]
    let maxSeconds: Int32
    let xLabels: [String]
    
    private let pointSpacing: CGFloat = 50
    private let labelWidth: CGFloat = 40
    
    var body: some View {
        // Calculate nice Y-axis ticks (round numbers)
        let yTicks = SummaryLineChartVm.companion.calculateYAxisTicks(maxSeconds: maxSeconds, tickCount: 5)
        let adjustedMaxSeconds = yTicks.last?.int32Value ?? 3600
        
        let pointCount = xLabels.count
        let totalWidth = pointCount > 1 ? labelWidth + pointSpacing * CGFloat(pointCount - 1) : labelWidth
        
        VStack {
            // Y-axis labels and chart area
            HStack(alignment: .top, spacing: 0) {
                // Y-axis labels - use GeometryReader for precise positioning
                GeometryReader { geo in
                    let height = geo.size.height
                    ZStack(alignment: .topTrailing) {
                        ForEach(yTicks, id: \.self) { tickValue in
                            let fraction = CGFloat(tickValue.int32Value) / CGFloat(adjustedMaxSeconds)
                            // Position from bottom: fraction=0 at bottom, fraction=1 at top
                            let offsetFromTop = height * (1 - fraction)
                            Text(SummaryLineChartVm.companion.formatSeconds(seconds: tickValue.int32Value))
                                .foregroundColor(.secondary)
                                .font(.system(size: 10, weight: .light))
                                .position(x: 20, y: offsetFromTop)
                        }
                    }
                }
                .frame(width: 45)
                
                // Chart canvas
                GeometryReader { geometry in
                    ScrollViewReader { scrollProxy in
                        ScrollView(.horizontal, showsIndicators: false) {
                            VStack(spacing: 4) {
                                // Chart canvas
                                Canvas { context, size in
                                    let canvasHeight = size.height
                                    let canvasWidth = size.width
                                    
                                    guard pointCount > 0 else { return }
                                    
                                    // Draw horizontal grid lines at each Y-axis tick
                                    let gridColor = Color.gray.opacity(0.3)
                                    for tickValue in yTicks {
                                        let y = canvasHeight - (CGFloat(tickValue.int32Value) / CGFloat(adjustedMaxSeconds) * canvasHeight)
                                        var path = Path()
                                        path.move(to: CGPoint(x: 0, y: y))
                                        path.addLine(to: CGPoint(x: canvasWidth, y: y))
                                        context.stroke(path, with: .color(gridColor), lineWidth: 1)
                                    }
                                    
                                    // Draw lines for each goal
                                    for lineData in linesData {
                                        guard !lineData.points.isEmpty else { continue }
                                        
                                        let color = lineData.color.toColor()
                                        var linePath = Path()
                                        
                                        for (index, point) in lineData.points.enumerated() {
                                            let x = labelWidth / 2 + pointSpacing * CGFloat(index)
                                            let y = canvasHeight - (CGFloat(point.seconds) / CGFloat(adjustedMaxSeconds) * canvasHeight)
                                            
                                            if index == 0 {
                                                linePath.move(to: CGPoint(x: x, y: y))
                                            } else {
                                                linePath.addLine(to: CGPoint(x: x, y: y))
                                            }
                                        }
                                        
                                        // Draw the line
                                        context.stroke(linePath, with: .color(color), style: StrokeStyle(lineWidth: 3, lineCap: .round, lineJoin: .round))
                                        
                                        // Draw points
                                        for (index, point) in lineData.points.enumerated() {
                                            let x = labelWidth / 2 + pointSpacing * CGFloat(index)
                                            let y = canvasHeight - (CGFloat(point.seconds) / CGFloat(adjustedMaxSeconds) * canvasHeight)
                                            
                                            let pointPath = Path(ellipseIn: CGRect(x: x - 4, y: y - 4, width: 8, height: 8))
                                            context.fill(pointPath, with: .color(color))
                                        }
                                    }
                                }
                                .frame(width: totalWidth, height: geometry.size.height - 20)
                                
                                // X-axis labels (inside same ScrollView for sync)
                                HStack(spacing: 0) {
                                    ForEach(Array(xLabels.enumerated()), id: \.offset) { index, label in
                                        Text(label)
                                            .font(.system(size: 9, weight: .light))
                                            .foregroundColor(.secondary)
                                            .frame(width: labelWidth)
                                        
                                        if index < xLabels.count - 1 {
                                            Spacer()
                                                .frame(width: pointSpacing - labelWidth)
                                        }
                                    }
                                }
                                .frame(width: totalWidth)
                            }
                        }
                    }
                }
            }
        }
    }
}

private struct GoalSelectionList: View {
    
    let selections: [SummaryLineChartVm.GoalSelectionUi]
    let onToggle: (Int32) -> Void
    
    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(selections, id: \.goalDb.id) { selection in
                    GoalSelectionItem(
                        selection: selection,
                        onToggle: { onToggle(selection.goalDb.id) }
                    )
                }
            }
            .padding(.horizontal, 8)
        }
    }
}

private struct GoalSelectionItem: View {
    
    let selection: SummaryLineChartVm.GoalSelectionUi
    let onToggle: () -> Void
    
    var body: some View {
        let backgroundColor = selection.isSelected
            ? selection.color.toColor().opacity(0.3)
            : Color(.systemGray5)
        
        Button(action: onToggle) {
            HStack {
                Circle()
                    .fill(selection.color.toColor())
                    .frame(width: 12, height: 12)
                
                Text(selection.title)
                    .font(.system(size: 13, weight: selection.isSelected ? .medium : .regular))
                    .foregroundColor(Color(.label))
                    .lineLimit(1)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(
                RoundedRectangle(cornerRadius: 8, style: .continuous)
                    .fill(backgroundColor)
            )
        }
    }
}

File: android_app/src/main/java/me/timeto/app/ui/summary/SummaryLineChartView.kt
package me.timeto.app.ui.summary

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.toColor
import me.timeto.app.ui.HStack
import me.timeto.app.ui.Screen
import me.timeto.app.ui.VStack
import me.timeto.app.ui.ZStack
import me.timeto.app.ui.c
import me.timeto.app.ui.rememberVm
import me.timeto.app.ui.roundedShape
import me.timeto.app.ui.squircleShape
import me.timeto.shared.DayBarsUi
import me.timeto.shared.vm.summary.SummaryLineChartVm
import me.timeto.shared.vm.summary.SummaryVm

@Composable
fun SummaryLineChartView(
    goalsUi: List<SummaryVm.GoalUi>,
    daysBarsUi: List<DayBarsUi>,
) {
    val (vm, state) = rememberVm(goalsUi, daysBarsUi) {
        SummaryLineChartVm(goalsUi, daysBarsUi)
    }

    Screen {
        VStack(
            modifier = Modifier
                .fillMaxSize()
                .background(c.bg)
                .verticalScroll(rememberScrollState()),
        ) {
            // Time unit selector
            TimeUnitSelector(
                currentUnit = state.timeUnit,
                onUnitSelected = { vm.setTimeUnit(it) }
            )

            // Line Chart
            if (state.linesData.isNotEmpty()) {
                LineChart(
                    linesData = state.linesData,
                    maxSeconds = state.maxSecondsOverall,
                    xLabels = state.xLabels,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(220.dp)
                        .padding(horizontal = 8.dp, vertical = 16.dp)
                )
            } else {
                ZStack(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(200.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = "Select activities to display",
                        color = c.secondaryText,
                        fontSize = 14.sp,
                    )
                }
            }

            // Goal selection
            Text(
                text = "Select Activities",
                modifier = Modifier
                    .padding(horizontal = 16.dp, vertical = 8.dp),
                color = c.text,
                fontSize = 14.sp,
                fontWeight = FontWeight.Medium,
            )

            GoalSelectionList(
                selections = state.goalSelections,
                onToggle = { goalId -> vm.toggleGoalSelection(goalId) }
            )

            Spacer(modifier = Modifier.height(20.dp))
        }
    }
}

@Composable
private fun TimeUnitSelector(
    currentUnit: SummaryLineChartVm.TimeUnit,
    onUnitSelected: (SummaryLineChartVm.TimeUnit) -> Unit,
) {
    HStack(
        modifier = Modifier
            .padding(horizontal = 16.dp, vertical = 12.dp)
            .fillMaxWidth(),
        horizontalArrangement = Arrangement.Center,
    ) {
        SummaryLineChartVm.TimeUnit.entries.forEach { unit ->
            val isSelected = unit == currentUnit
            Text(
                text = unit.title,
                modifier = Modifier
                    .clip(squircleShape)
                    .background(if (isSelected) c.blue else c.transparent)
                    .clickable { onUnitSelected(unit) }
                    .padding(horizontal = 16.dp, vertical = 8.dp),
                color = if (isSelected) c.white else c.text,
                fontSize = 14.sp,
                fontWeight = if (isSelected) FontWeight.Medium else FontWeight.Normal,
            )
        }
    }
}

@Composable
private fun LineChart(
    linesData: List<SummaryLineChartVm.LineData>,
    maxSeconds: Int,
    xLabels: List<String>,
    modifier: Modifier = Modifier,
) {
    // Calculate nice Y-axis ticks (round numbers)
    val yTicks = SummaryLineChartVm.calculateYAxisTicks(maxSeconds)
    val adjustedMaxSeconds = yTicks.lastOrNull() ?: 3600
    
    val scrollState = rememberScrollState()
    val pointSpacingDp = 50.dp
    val labelWidthDp = 40.dp
    val pointCount = xLabels.size
    
    // Calculate total width: first point at labelWidth/2, then spacing between points
    val totalWidthDp = if (pointCount > 1) {
        labelWidthDp + pointSpacingDp * (pointCount - 1)
    } else {
        labelWidthDp
    }

    Column(modifier = modifier) {
        // Y-axis labels and chart area
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
        ) {
            // Y-axis labels - use BoxWithConstraints to position labels precisely
            BoxWithConstraints(
                modifier = Modifier
                    .width(45.dp)
                    .fillMaxHeight()
            ) {
                val boxHeight = maxHeight
                yTicks.forEach { tickValue ->
                    val fraction = tickValue.toFloat() / adjustedMaxSeconds
                    // Position from bottom: fraction=0 at bottom, fraction=1 at top
                    val offsetFromBottom = boxHeight * fraction
                    Text(
                        text = SummaryLineChartVm.formatSeconds(tickValue),
                        modifier = Modifier
                            .align(Alignment.BottomEnd)
                            .offset(y = -offsetFromBottom + 5.dp) // +5.dp to center text on line
                            .padding(end = 4.dp),
                        color = c.secondaryText,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Light,
                    )
                }
            }

            // Chart canvas with horizontal scroll
            Box(
                modifier = Modifier
                    .weight(1f)
                    .fillMaxHeight()
                    .horizontalScroll(scrollState)
            ) {
                Canvas(
                    modifier = Modifier
                        .width(totalWidthDp)
                        .fillMaxHeight()
                ) {
                    val canvasWidth = size.width
                    val canvasHeight = size.height
                    val labelWidth = labelWidthDp.toPx()
                    val pointSpacing = pointSpacingDp.toPx()

                    if (pointCount == 0) return@Canvas

                    // Draw horizontal grid lines at each Y-axis tick
                    val gridColor = Color.Gray.copy(alpha = 0.5f)
                    yTicks.forEach { tickValue ->
                        val y = canvasHeight - (tickValue.toFloat() / adjustedMaxSeconds * canvasHeight)
                        drawLine(
                            color = gridColor,
                            start = Offset(0f, y),
                            end = Offset(canvasWidth, y),
                            strokeWidth = 1f,
                        )
                    }

                    // Draw lines for each goal
                    linesData.forEach { lineData ->
                        if (lineData.points.isEmpty()) return@forEach

                        val path = Path()
                        val color = lineData.color.toColor()

                        lineData.points.forEachIndexed { index, point ->
                            // First point at labelWidth/2, then spaced by pointSpacing
                            val x = labelWidth / 2 + pointSpacing * index
                            val y = canvasHeight - (point.seconds.toFloat() / adjustedMaxSeconds * canvasHeight)

                            if (index == 0) {
                                path.moveTo(x, y)
                            } else {
                                path.lineTo(x, y)
                            }
                        }

                        // Draw the line
                        drawPath(
                            path = path,
                            color = color,
                            style = Stroke(
                                width = 3f,
                                cap = StrokeCap.Round,
                                join = StrokeJoin.Round,
                            )
                        )

                        // Draw points
                        lineData.points.forEachIndexed { index, point ->
                            val x = labelWidth / 2 + pointSpacing * index
                            val y = canvasHeight - (point.seconds.toFloat() / adjustedMaxSeconds * canvasHeight)

                            drawCircle(
                                color = color,
                                radius = 5f,
                                center = Offset(x, y),
                            )
                        }
                    }
                }
            }
        }

        // X-axis labels (shared scrollState for sync)
        if (xLabels.isNotEmpty()) {
            Row(
                modifier = Modifier
                    .padding(start = 45.dp)
                    .horizontalScroll(scrollState)
                    .padding(top = 4.dp),
            ) {
                xLabels.forEachIndexed { index, label ->
                    Text(
                        text = label,
                        modifier = Modifier.width(labelWidthDp),
                        color = c.secondaryText,
                        fontSize = 9.sp,
                        fontWeight = FontWeight.Light,
                        textAlign = TextAlign.Center,
                        maxLines = 1,
                    )
                    if (index < xLabels.size - 1) {
                        Spacer(modifier = Modifier.width(pointSpacingDp - labelWidthDp))
                    }
                }
            }
        }
    }
}

@Composable
private fun GoalSelectionList(
    selections: List<SummaryLineChartVm.GoalSelectionUi>,
    onToggle: (Int) -> Unit,
) {
    LazyRow(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        items(selections, key = { it.goalDb.id }) { selection ->
            GoalSelectionItem(
                selection = selection,
                onToggle = { onToggle(selection.goalDb.id) }
            )
        }
    }
}

@Composable
private fun GoalSelectionItem(
    selection: SummaryLineChartVm.GoalSelectionUi,
    onToggle: () -> Unit,
) {
    val backgroundColor = if (selection.isSelected)
        selection.color.toColor().copy(alpha = 0.3f)
    else
        c.gray5

    val borderWidth = animateDpAsState(
        targetValue = if (selection.isSelected) 2.dp else 0.dp,
        animationSpec = spring(stiffness = Spring.StiffnessLow),
        label = "borderWidth"
    )

    HStack(
        modifier = Modifier
            .clip(roundedShape)
            .background(backgroundColor)
            .then(
                if (selection.isSelected) {
                    Modifier.background(
                        color = selection.color.toColor().copy(alpha = 0.15f),
                        shape = roundedShape
                    )
                } else Modifier
            )
            .clickable { onToggle() }
            .padding(horizontal = 12.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier = Modifier
                .size(12.dp)
                .clip(RoundedCornerShape(6.dp))
                .background(selection.color.toColor())
        )

        Text(
            text = selection.title,
            modifier = Modifier
                .padding(start = 8.dp),
            color = c.text,
            fontSize = 13.sp,
            fontWeight = if (selection.isSelected) FontWeight.Medium else FontWeight.Normal,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}
